src/compiler/scala/tools/nsc/transform/UnCurry.scala

```diff
@@ -65,19 +66,37 @@ abstract class UnCurry extends InfoTransform
 // uncurry and uncurryType expand type aliases
 
   class UnCurryTransformer(unit: CompilationUnit) extends TypingTransformer(unit) {
-    private val inlineFunctionExpansion = settings.Ydelambdafy.value == "inline"
+    private val forceExpandFunction = settings.Ydelambdafy.value == "inline"
     private var needTryLift       = false
     private var inConstructorFlag = 0L
     private val byNameArgs        = mutable.HashSet[Tree]()
     private val noApply           = mutable.HashSet[Tree]()
     private val newMembers        = mutable.Map[Symbol, mutable.Buffer[Tree]]()
 
-    private lazy val forceSpecializationInfoTransformOfFunctionN: Unit = {
-      if (currentRun.specializePhase != NoPhase) { // be robust in case of -Ystop-after:uncurry
-        exitingSpecialize {
-          FunctionClass.seq.foreach(cls => cls.info)
-        }
-      }
+    // Expand `Function`s in constructors to class instance creation (SI-6666, SI-8363)
+    // We use Java's LambdaMetaFactory (LMF), which requires an interface for the sam's owner
+    private def mustExpandFunction(fun: Function) = forceExpandFunction || {
+      // (TODO: Can't use isInterface, yet, as it hasn't been updated for the new trait encoding)
+      val canUseLambdaMetaFactory = inConstructorFlag == 0 && (fun.attachments.get[SAMFunction].map(_.samTp) match {
+        case Some(userDefinedSamTp) =>
+          val tpSym = erasure.javaErasure(userDefinedSamTp).typeSymbol // we only care about what ends up in the bytecode
+          (
+            // LMF only supports interfaces
+            (tpSym.isJavaInterface || tpSym.isTrait)
+            // Unless tpSym.isStatic, even if the constructor is zero-argument now, it may acquire arguments in explicit outer or lambdalift.
+            // This is an impl restriction to simplify the decision of whether to expand the SAM during uncurry
+            // (when we don't yet know whether it will receive an outer pointer in explicit outer or whether lambda lift will add proxies for captures).
+            // When we delay sam expansion until after explicit outer & lambda lift, we could decide there whether
+            // to expand sam at compile time or use LMF, and this implementation restriction could be lifted.
+            && tpSym.isStatic
+            // impl restriction -- we currently use the boxed apply, so not really useful to allow specialized sam types (https://github.com/scala/scala/pull/4971#issuecomment-198119167)
+            && !tpSym.isSpecialized
```

Comment:
How about 
```
def memberBecomesSpecialized(sym: Symbol) = needsSpecialization(typeEnv(sym.owner), sym)
```
//cc @dragos, since we've caught your eye